platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/InternalProblemSpec.java

```diff
@@ -24,15 +25,21 @@
 
 public interface InternalProblemSpec extends ProblemSpec {
 
+
     /**
-     * Specifies arbitrary data associated with this problem.
+     * Attaches additional data describing the problem.
+     * <p>
+     * Only the types listed for {@link AdditionalData} can be used as arguments, otherwise an invalid problem report will be created.
      * <p>
-     * The only supported value type is {@link String}. Future Gradle versions may support additional types.
+     * If not additional data was configured for this problem, then a new instance will be created. If additional data was already configured, then the existing instance will be used and the configuration will be applied to it.
      *
+     * @param type the type of the additional data (see the AdditionalData interface for the list of supported types)
+     * @param config  The action configuring the additional data
      * @return this
-     * @throws RuntimeException for null values and for values with unsupported type.
+     * @param <T> The type of the additional data
+     * @param <U> The type of the configurator object that will be applied to the additional data
      */
-    InternalProblemSpec additionalData(String key, Object value);
+    <T extends AdditionalData<U>, U> InternalProblemSpec additionalData(Class<? extends AdditionalData<U>> type, Action<? super U> config);
```

Comment:
```suggestion
    <T extends AdditionalData, U extends Builder<T>> InternalProblemSpec additionalData(Class<U> builderType, Action<? super U> config);
```

❓ Would something like this work? So you'd need to specify the builder type when adding additional data instead of specifying the builder with the additional data itself. That way consumers don't need to worry about the builder type.

Reply:
That's fair, I'll take a stab at it.

Reply:
The nice thing about the current implementation is the ease of use of the API. For example, when you write something like

     .additionalData(DeprecationData.class,

The type system will only allow you to pass an object of type `Action<DeprecationDataSpec>` as a second parameter. 

> That way consumers don't need to worry about the builder type.

It's not exactly a builder, but a spec type; a companion interface declaring what are the allowed configuration options for particular additional data type. Is this something that you'd disallow seeing on the consumer-side?

Reply:
I think what you'd pass in the API is the spec type, and then the spec type should know the additional data it can build. The consumer should only be interested in the actual data, not what was used to construct it, right?